### PR TITLE
Add advertise support to sync client

### LINKED ISSUE #16: Add advertise support to sync client.
Not quite sure how I want to expose this... potentially I'll add an advertise method that delegates to `async_client`.

### PR DESCRIPTION
**Pull request by breerly**
closes #16 

Still need to add some additional tests & docs.

### CODE REVIEW

**Comment by breerly** [comment]
cc @abhinav @junchaowu @blampe

File: tests/sync/test_client.py [hunk 1]
```diff
@@ -43,3 +45,30 @@ def test_sync_client_should_get_raw_response(tchannel_server):
...
+    result = client.advertise(routers)
+
+    assert result.header == ""
+    # @todo https://github.com/uber/tchannel-python/issues/34
+    assert result.body == json.dumps(body)
+    assert client.async_client.peers.hosts == routers
+
+
+def test_advertise_exception_should_propagate(tchannel_server):
+    pass
```
Comment (abhinav) [inline tests/sync/test_client.py]: ?

File: tchannel/sync/client.py [hunk 2]
```diff
@@ -85,6 +85,27 @@ def request(self, *args, **kwargs):
 
         return operation
 
+    def advertise(self, routers, name=None):
+
```
Comment (abhinav) [inline tchannel/sync/client.py]: docs

**Comment by breerly** [comment]
Think I found a bug in threadloop... if I set the timeout to 30seconds, and then try and kill the program with ^C, it doesn't work.

**Comment by breerly** [comment]
Filed an issue - will fix it tomorrow: https://github.com/breerly/threadloop/issues/24

**Comment by blampe** [comment]
:+1:

File: tchannel/tornado/hyperbahn.py [hunk 3]
```diff
@@ -84,7 +84,7 @@ def _advertise_with_backoff(tchannel, service, timeout=None):
     start = time.time()
 
     while True:
-        if timeout and time.time() - start > 30:
```
Comment (breerly) [inline tchannel/tornado/hyperbahn.py]: by the way this was a bug

File: tchannel/sync/client.py [hunk 4]
```diff
@@ -85,6 +85,29 @@ def request(self, *args, **kwargs):
...
+            header = yield response.get_header()
+            body = yield response.get_body()
+
+            result = Response(header, body)
+
+            raise gen.Return(result)
+
+        future = self.threadloop.submit(make_request)
+        result = future.result()
```
Comment (breerly) [inline tchannel/sync/client.py]: Need timeout here.

File: tchannel/sync/client.py [hunk 5]
```diff
@@ -65,6 +67,7 @@ def __init__(self, name, process_name=None, known_peers=None, trace=False):
         )
         self.threadloop = ThreadLoop()
         self.threadloop.start()
+        self.wait_on_advertise_until = 1
```
Comment (junchaowu) [inline tchannel/sync/client.py]: where is it used?

File: tchannel/sync/client.py [hunk 6]
```diff
@@ -85,6 +88,43 @@ def request(self, *args, **kwargs):
...
+
+            result = Response(header, body)
+
+            raise gen.Return(result)
+
+        future = self.threadloop.submit(make_request)
+
+        # we're going to wait 1s longer than advertises
+        # timeout mechanism, so it has a chance to timeout
+        wait_until = timeout or FIRST_ADVERTISE_TIME
```
Comment (abhinav) [inline tchannel/sync/client.py]: If you're depending on the default value for `FIRST_ADVERTISE_TIME` anyway, might as well use it as the default value for `timeout` before passing it down to the async client.

**Comment by junchaowu** [comment]
Would be great to write `best practice` of sync_client either in doc string or guide

File: tchannel/sync/client.py [hunk 7]
```diff
@@ -85,6 +88,43 @@ def request(self, *args, **kwargs):
...
+        wait_until += 1
+
+        # block for advertise's first response,
+        # using wait_until as a fallback timeout mechanism
+        try:
+            result = future.result(wait_until)
+        except TimeoutError:
+            raise AdvertiseError(
+                "Failed to register with Hyperbahn "
+                "(advertise did not timeout in time)"
```
Comment (abhinav) [inline tchannel/sync/client.py]: "advertise did not time out in time" doesn't make much sense. Just mention that advertise timed out.

Comment (breerly) [inline tchannel/sync/client.py]: @abhinav I want a tiny way to know that this exception is different than the other actual advertise error.

File: tchannel/sync/client.py (hunk 6 above)
Comment (breerly) [inline tchannel/sync/client.py]: It's cool - I want to keep the method signatures simple for potential other wrapping functionality.

File: tchannel/sync/client.py (hunk 5 above)
Comment (breerly) [inline tchannel/sync/client.py]: good catch; deleting

**Comment by breerly** [comment]
fixes #36

File: CHANGES.rst [hunk 8]
```diff
@@ -1,10 +1,11 @@
 Changelog
 =========
 
-0.10.4 (unreleased)
+0.11.0 (unreleased)
 -------------------
 
-- Nothing changed yet.
+- Added advertise support to sync client in ``tchannel.sync.TChannelSyncClient.advertise``.
+- **BREAKING**  - renamed ``router`` argument to ``routers`` in ``tchannel.tornado.TChannel.advertise``.
```
Comment (blampe) [inline CHANGES.rst]: Nit: two spaces :)

**Comment by blampe** [comment]
:shipit:

File: CHANGES.rst (hunk 8 above)
Comment (breerly) [inline CHANGES.rst]: @blampe huh?

**Pull request merged**